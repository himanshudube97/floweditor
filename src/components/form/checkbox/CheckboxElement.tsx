import { react as bindCallbacks } from 'auto-bind';
import { FormElementProps } from 'components/form/FormElement';
import * as React from 'react';

export interface CheckboxElementProps extends FormElementProps {
  checked: boolean;
  title?: string;
  description?: string;
  onChange?(checked: boolean): void;
}

interface CheckboxState {
  checked: boolean;
}

export const boxIco = 'fe-square';
export const checkedBoxIco = 'fe-check-square';

export const checkboxSpecId = 'checkbox';
export const titleSpecId = 'title';
export const descSpecId = 'description';

export default class CheckboxElement extends React.Component<CheckboxElementProps, CheckboxState> {
  private checkbox: HTMLElement;
  constructor(props: any) {
    super(props);

    this.state = {
      checked: this.props.checked
    };

    bindCallbacks(this, {
      include: [/^handle/]
    });
  }

  public handleChange(event: any) {
    if (event) {
      event.preventDefault();
      event.stopPropagation();
    }
    if (this.checkbox) {
      const checked = (this.checkbox as any).checked;
      if (checked !== this.state.checked && checked !== undefined) {
        this.setState({ checked }, () => {
          if (this.props.onChange) {
            this.props.onChange(checked);
          }
        });
      }
    }
  }

  public componentDidMount(): void {
    if (this.props.checked) {
      this.setState({ checked: this.props.checked });
    }
    if (this.checkbox) {
      this.checkbox.addEventListener('change', this.handleChange);
    }
  }

  public componentWillUnmount(): void {
    this.checkbox.removeEventListener('change', this.handleChange);
  }

  /* istanbul ignore next */
  public validate(): boolean {
    return true;
  }

  public render(): JSX.Element {
    const optional: any = {};
    if (this.state.checked) {
      optional['checked'] = true;
    }
    return (
      <temba-checkbox
        ref={(ele: any) => {
          this.checkbox = ele;
        }}
        {...optional}
        label={this.props.title}
        help_text={this.props.description}
      ></temba-checkbox>
    );
  }
}
