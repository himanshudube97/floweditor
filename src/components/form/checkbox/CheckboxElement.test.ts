import { composeComponentTestUtils } from 'testUtils';

import CheckboxElement, { CheckboxElementProps } from './CheckboxElement';

const baseProps: CheckboxElementProps = {
  name: 'Checkbox',
  checked: false
};

const { setup } = composeComponentTestUtils(CheckboxElement, baseProps);

describe(CheckboxElement.name, () => {
  it('should render a checkbox element with title, description', () => {
    const { wrapper } = setup(false, {
      $merge: {
        title: 'Checkbox',
        description: 'All Destinations',
        labelClassName: 'label',
        checkboxClassName: 'checkbox',
        onChange: jest.fn()
      }
    });
    expect(wrapper).toMatchSnapshot('unchecked');
  });
});
