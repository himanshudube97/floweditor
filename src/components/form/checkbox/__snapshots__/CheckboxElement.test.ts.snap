// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxElement should render a checkbox element with title, description: unchecked 1`] = `
<CheckboxElement
  checkboxClassName="checkbox"
  checked={false}
  description="All Destinations"
  labelClassName="label"
  name="Checkbox"
  onChange={[MockFunction]}
  title="Checkbox"
>
  <temba-checkbox
    help_text="All Destinations"
    label="Checkbox"
  />
</CheckboxElement>
`;
